gui/gui/src/TGWindow.cxx

```diff
@@ -51,6 +53,13 @@ TGWindow::TGWindow(const TGWindow *p, Int_t x, Int_t y, UInt_t w, UInt_t h,
    fParent = 0;
    fNeedRedraw = kFALSE;
 
+   if (!p && !gClient && !gApplication) {
+      ::Error("TGWindow::TGWindow",
+              "gClient and gApplication are nullptr!\n"
+              "Please add a TApplication instance in the main() function of your application\n");
+      std::exit(EXIT_FAILURE);
```

Comment:
Why not use `::Fatal` ?

Reply:
> Why not use `::Fatal` ?

I tried, but it crashes (at least on Windows), but this should never happen anyway...

Reply:
What do you mean by 'crash' :) ?   Fatal is suppose to exit the process.  Does Fatal not work just here or in general on Windows?

Reply:
You know, it's the usual GUI stuff with the Window Manager. Do you need me to reproduce and provide the complete stack trace (if any)?

Reply:
This sounds like TROOT::Terminate is not 'waiting' for the GUI to properly tear down.   So yes, **if** it is not time consuming to step, a reproducer + stack traces will be helpful.

Reply:
Working on it...

Reply:
Here is the stack trace:
```
aborting

==========================================
=============== STACKTRACE ===============
==========================================


================ Thread 0 ================
  C:\Users\bellenot\git\master\core\foundation\src\TError.cxx(152): ErrorHandler()
  C:\Users\bellenot\git\master\core\foundation\src\TError.cxx(249): Fatal()
  C:\Users\bellenot\git\master\gui\gui\src\TGWindow.cxx(57): TGWindow::TGWindow()
  C:\Users\bellenot\git\master\gui\gui\src\TGFrame.cxx(147): TGFrame::TGFrame()
  C:\Users\bellenot\git\master\gui\gui\src\TGFrame.cxx(847): TGCompositeFrame::TGCompositeFrame()
  C:\Users\bellenot\git\master\gui\gui\src\TGFrame.cxx(1430): TGMainFrame::TGMainFrame()
  C:\Users\bellenot\git\master\tree\treeviewer\src\TTreeViewer.cxx(346): TTreeViewer::TTreeViewer()
  0x2c30ae ??
  C:\Users\bellenot\git\master\core\metacling\src\TClingCallFunc.cxx(1842): TClingCallFunc::exec()
  C:\Users\bellenot\git\master\core\metacling\src\TClingCallFunc.cxx(2042): <lambda_45b0f5d8ad983e3738d89b1979c56f47>::operator()()
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\type_traits(1579): std::_Invoker_functor::_Call<<lambda_45b0f5d8ad983e3738d89b1979c56f47> &,void *,cling::Value &>()
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\type_traits(1579): std::invoke<<lambda_45b0f5d8ad983e3738d89b1979c56f47> &,void *,cling::Value &>()
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\type_traits(1597): std::_Invoker_ret<void,1>::_Call<<lambda_45b0f5d8ad983e3738d89b1979c56f47> &,void *,cling::Value &>()
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\functional(926): std::_Func_impl_no_alloc<<lambda_45b0f5d8ad983e3738d89b1979c56f47>,void,void *,cling::Value &>::_Do_call()
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\functional(970): std::_Func_class<void,void *,cling::Value &>::operator()()
  C:\Users\bellenot\git\master\core\metacling\src\TClingCallFunc.cxx(2065): TClingCallFunc::exec_with_valref_return()
  C:\Users\bellenot\git\master\core\metacling\src\TClingCallFunc.cxx(2119): TClingCallFunc::ExecT<long>()
  C:\Users\bellenot\git\master\core\metacling\src\TClingCallFunc.cxx(2132): TClingCallFunc::ExecInt()
  C:\Users\bellenot\git\master\core\metacling\src\TCling.cxx(7742): TCling::CallFunc_ExecInt()
  C:\Users\bellenot\git\master\core\meta\src\TMethodCall.cxx(457): TMethodCall::Execute()
  C:\Users\bellenot\git\master\core\meta\inc\TMethodCall.h(136): TMethodCall::Execute()
  C:\Users\bellenot\git\master\core\base\inc\TPluginManager.h(162): TPluginHandler::ExecPluginImpl<TTree *>()
  C:\Users\bellenot\git\master\core\base\inc\TPluginManager.h(173): TPluginHandler::ExecPlugin<TTree *>()
  C:\Users\bellenot\git\master\tree\treeplayer\src\TTreePlayer.cxx(2969): TTreePlayer::StartViewer()
  C:\Users\bellenot\git\master\tree\tree\src\TTree.cxx(9264): TTree::StartViewer()
  C:\Users\bellenot\rootdev\test_treeviewer.cxx(22): main()
  d:\agent\_work\6\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(288): __scrt_common_main_seh()
  KERNEL32!BaseThreadInitThunk()
  ntdll!RtlGetAppContainerNamedObjectPath()
  ntdll!RtlGetAppContainerNamedObjectPath()

================ Thread 1 ================
  ntdll!ZwWaitForWorkViaWorkerFactory()
  KERNEL32!BaseThreadInitThunk()
  ntdll!RtlGetAppContainerNamedObjectPath()
  ntdll!RtlGetAppContainerNamedObjectPath()

================ Thread 2 ================
  ntdll!ZwWaitForWorkViaWorkerFactory()
  KERNEL32!BaseThreadInitThunk()
  ntdll!RtlGetAppContainerNamedObjectPath()
  ntdll!RtlGetAppContainerNamedObjectPath()

================ Thread 3 ================
  ntdll!ZwWaitForWorkViaWorkerFactory()
  KERNEL32!BaseThreadInitThunk()
  ntdll!RtlGetAppContainerNamedObjectPath()
  ntdll!RtlGetAppContainerNamedObjectPath()

================ Thread 4 ================
  ntdll!NtDelayExecution()
  KERNELBASE!Sleep()
  C:\Users\bellenot\git\master\core\winnt\src\TWinNTSystem.cxx(4377): TWinNTSystem::TimerThread()
  C:\Users\bellenot\git\master\core\winnt\inc\TWinNTSystem.h(99): TWinNTSystem::ThreadStub()
  KERNEL32!BaseThreadInitThunk()
  ntdll!RtlGetAppContainerNamedObjectPath()
  ntdll!RtlGetAppContainerNamedObjectPath()

================ Thread 5 ================
  win32u!NtUserGetMessage()
  C:\Users\bellenot\git\master\core\winnt\src\TWinNTSystem.cxx(496): `anonymous namespace'::GUIThreadMessageProcessingLoop()
  KERNEL32!BaseThreadInitThunk()
  ntdll!RtlGetAppContainerNamedObjectPath()
  ntdll!RtlGetAppContainerNamedObjectPath()

==========================================
============= END STACKTRACE =============
==========================================
```

Reply:
The intended behavior of Fatal (you can run it on linux to see), i.e. abort and print a stack trace.
Now we could discuss whether ::Fatal should close down the other thread first.

Reply:
If you want to avoid the stack trace.  Then rather that std::exit, I would prefer a call to Terminate or Exit.

Reply:
OK, I'll investigate how to make it working on Windows and Linux. Going to take quite some time...

Reply:
I am confused.  ::Fatal already working as expected and alternatively replacing std::exit with gSystem->Exit should be straight-forward.  What am I missing? (I.e. what do you mean to investigate?)

Reply:
Sorry, I was referring to this:
> The intended behavior of Fatal (you can run it on linux to see), i.e. abort and print a stack trace.
> Now we could discuss whether ::Fatal should close down the other thread first.

And about this statement
> I am confused. ::Fatal already working as expected and alternatively replacing std::exit with gSystem->Exit should be straight-forward. What am I missing? (I.e. what do you mean to investigate?)
`::Fatal` doesn't work, it produces the stack trace I posted. And I'll try `gSystem->Exit`. It's what I mean by "I'll investigate".

Reply:
ok

Reply:
@pcanal ok, using `gSystem->Exit()` works. Please let me know if it's good enough for you now.